# Assignment: Data Wrangling
## `! git clone https://github.com/DS3001/wrangling`
## Do Q2, and one of Q1 or Q3.

**Q1.** Open the "tidy_data.pdf" document in the repo, which is a paper called Tidy Data by Hadley Wickham.

  1. Read the abstract. What is this paper about?
  2. Read the introduction. What is the "tidy data standard" intended to accomplish?
  3. Read the intro to section 2. What does this sentence mean: "Like families, tidy datasets are all alike but every messy dataset is messy in its own way." What does this sentence mean: "For a given dataset, it’s usually easy to figure out what are observations and what are variables, but it is surprisingly difficult to precisely define variables and observations in general."
  4. Read Section 2.2. How does Wickham define values, variables, and observations?
  5. How is "Tidy Data" defined in section 2.3?
  6. Read the intro to Section 3 and Section 3.1. What are the 5 most common problems with messy datasets? Why are the data in Table 4 messy? What is "melting" a dataset?
  7. Why, specifically, is table 11 messy but table 12 tidy and "molten"?
  8. Read Section 6. What is the "chicken-and-egg" problem with focusing on tidy data? What does Wickham hope happens in the future with further work on the subject of data wrangling?

1) This paper is about how it is often emphasized that cleaning data is a crucial part of data analysis, however, there is a lack of an examination of the various data cleaning techniques and which ones are more effective than others. Through this paper, Wickham seeks to lay out an analysis of a part data cleaning, data tidying, and why data tidying ensures that datasets are easier to work with. By having the structure that variables are colums, observations are rows, and observational units are tables, this makes it easier to also come up with tools that enable easier data tidying.

2) The "tidy data standard" seeks to create a method of data exploration and analysis that is streamlined and easy to work with, facilitating a focus on the problem at hand rather than spending too much time on tidying the data. Addtionally, it intends to find a way to create data analysis tools in a manner that they can work with each other without having to transform the data in order to input it into another data analysis tool, as seen with other methods.

3) The first sentence refers to how each tidy dataset follows certain guidelines so that tidy datasets tend to have similar organizational patterns and are composed of similar characteristics. Conversly, data that is messy are composed of a variety of problems that are specific to that dataset. Families follow a similar structure, as happier familiers tend to have certain of the same characteristics whereas unhappy families may arise due to a multitude of issues.

The second sentence refers to how the layout of a dataset enables one to understand broadly that observations will be in the rows and variables will be in the columns. However, the problem arise where although you may be able to locate them, one may not acutually understand what that obersvation or variable means in the context of the dataset. Addtionally, one may not understand what entails an observation or row or what factors contributed to the labeling of a column or the documentation of an observation.

4) Wickham defines a value as numbers for quantitative data or strings for qualitative data. A collection of values is what composed a dataset. Within that dataset are variables, which entail values that all measure the same element or feature through units. An observation is made up of all values that have the same units through which they are measured over certain attributes.

5) Tidy data is a standardized manner of displaying data where individual variables creates a column, an individual observation creates a row, and a table is created through every one type of observational unit. The arrangement of how rows, columns, and tables are matched with the observations are what determines if a dataset is messy or tidy.

6) The first most common problem with messy datasets is that column headers are labeled with values rather than variable names.
The second most common problem with messy datasets is that combining and storing multiple variables into one column. 
The third most common problem with messy datasets is that rows and columns both entails variables
The fourth most common problem is that a table will contain various types of observational units.
The fifth most common problem is that across multiple table, a single observational unit is kept.
The data is table 4 is messy because the column names are composed of the values of income. Addtionally, the question of interest is how income and religion are connected, which should include three variables - religion, which is correctly shown, income, which has values as a header, and frequency, which is not clearly shown with this table. 
Melting a dataset refers to stacking the data, or changing the columns into rows. This will allow for the values that are headers for columns to become rows.

7) Table 11 is messy because it has multiple columns that represent days, however, days should be a value. Addtionally, within the element variable are tmax and tmin. In order to store both the tmax and tmin for each day, the months end up doubling. This is messy because this means these variables are being stored as rows instead of columns. Table 12 is molten because it changed the columns of day values into rows. This also helped reduce columns by having the year, month, and day together as one under the date column. It is ultimatly made tidy by turning the tmax and tmin rows into columns. This makes the dataset tidy because now each column is a variable and the rows are actual values.

8) The "chicken and egg" problem with tidy data refers to how creating tidy data by utilizing specific tidy tools leads to those tools being soley connected to tidy data. This would make it harder to modify individual data structures or certain data tools without having to adhere to what has already been established through the reciporal connection of the construction of tidy data allows for more tools to be installed and it is these tools that give rise to tidy data.




**Q2.** This question provides some practice cleaning variables which have common problems.
1. Numeric variable: For `./data/airbnb_hw.csv`, clean the `Price` variable as well as you can, and explain the choices you make. How many missing values do you end up with? (Hint: What happens to the formatting when a price goes over 999 dollars, say from 675 to 1,112?)
2. Categorical variable: For the `./data/sharks.csv` data covered in the lecture, clean the "Type" variable as well as you can, and explain the choices you make.
3. Dummy variable: For the pretrial data covered in the lecture, clean the `WhetherDefendantWasReleasedPretrial` variable as well as you can, and, in particular, replace missing values with `np.nan`.
4. Missing values, not at random: For the pretrial data covered in the lecture, clean the `ImposedSentenceAllChargeInContactEvent` variable as well as you can, and explain the choices you make. (Hint: Look at the `SentenceTypeAllChargesAtConvictionInContactEvent` variable.)

In [231]:
# Number 1
import numpy as np
import pandas as pd

airbnb = pd.read_csv("/Users/borayadiul/Desktop/DS 3001/wrangling/assignment/data/airbnb_hw.csv", low_memory=False)
price1 = airbnb['Price']
print(price1. unique())
price2 = price1.str.replace(",", "")
print(price2.unique())
price3 = pd.to_numeric(price2, errors='coerce')
print(price3.unique())


['145' '37' '28' '199' '549' '149' '250' '90' '270' '290' '170' '59' '49'
 '68' '285' '75' '100' '150' '700' '125' '175' '40' '89' '95' '99' '499'
 '120' '79' '110' '180' '143' '230' '350' '135' '85' '60' '70' '55' '44'
 '200' '165' '115' '74' '84' '129' '50' '185' '80' '190' '140' '45' '65'
 '225' '600' '109' '1,990' '73' '240' '72' '105' '155' '160' '42' '132'
 '117' '295' '280' '159' '107' '69' '239' '220' '399' '130' '375' '585'
 '275' '139' '260' '35' '133' '300' '289' '179' '98' '195' '29' '27' '39'
 '249' '192' '142' '169' '1,000' '131' '138' '113' '122' '329' '101' '475'
 '238' '272' '308' '126' '235' '315' '248' '128' '56' '207' '450' '215'
 '210' '385' '445' '136' '247' '118' '77' '76' '92' '198' '205' '299'
 '222' '245' '104' '153' '349' '114' '320' '292' '226' '420' '500' '325'
 '307' '78' '265' '108' '123' '189' '32' '58' '86' '219' '800' '335' '63'
 '229' '425' '67' '87' '1,200' '158' '650' '234' '310' '695' '400' '166'
 '119' '62' '168' '340' '479' '43' '395' '144' '52' 

For this problem, I first imported the data using pd.read_csv, since the original file was a .csv file.
I then created a vector, price1, that only included the price vector in order to focus my cleaning on that vector.
After, I printed out the unique values within price1 in order to see this data and what needed to be cleaned.
I noticed that the values had single quotation marks, indicating that they were strings and addtionally, values over 999 had a comma for the thousands.
I decided to remove the commas as well as change the data type from string to numeric to make it easier to use this data for further analysis, and to make computing difference statistics cleaner. I assigned these changes to a new vector to ensure that I could go back to unedited data in case of a mistake.
I then used .unique() for the cleaned data to ensure that the commas and single quotation marks were gone.

In [230]:
# Number 2
import numpy as np
import pandas as pd
sharks = pd.read_csv("/Users/borayadiul/Desktop/DS 3001/wrangling/assignment/data/sharks.csv", low_memory=False)
type1 = sharks["Type"]
print(type1.value_counts())
print(sum(type1.isnull()))

type1 = type1.replace(['Watercraft', 'Boat','Boating','Boatomg'], "Vessel_related")
type1 = type1.replace(['Invalid', 'Questionable','Unconfirmed','Unverified','Under investigation'],"Needs_Examination")
clean_type1 = type1.dropna()
print(clean_type1)
print(sum(clean_type1.isnull()))
clean_type1.value_counts()


Type
Unprovoked             4716
Provoked                593
Invalid                 552
Sea Disaster            239
Watercraft              142
Boat                    109
Boating                  92
Questionable             10
Unconfirmed               1
Unverified                1
Under investigation       1
Boatomg                   1
Name: count, dtype: int64
5
0       Unprovoked
1         Provoked
2       Unprovoked
3       Unprovoked
4       Unprovoked
           ...    
6457    Unprovoked
6458    Unprovoked
6459    Unprovoked
6460    Unprovoked
6461    Unprovoked
Name: Type, Length: 6457, dtype: object
0


Type
Unprovoked           4716
Provoked              593
Needs_Examination     565
Vessel_related        344
Sea Disaster          239
Name: count, dtype: int64

For this question, I also imported the data using pd.read_csv. I then created a new vector that only included the type column for the sharks data in order to make cleaning easier and direct. I used value.counts() to see just how many times a unique value appears within the type vector. I did this to see if there were any categories that could be grouped together on the basis of similar characteristics of the incident, which would help with the overall analysis. Following that, I looked at if there were any null values that needed to be cleaned. After see that there were no null values, I then updated the type1 vector by grouping together Watercraft, Boat, Boating, and Boatomg as Vessel_related and replacing those from the original data. I did this because it seemed like all of these ones had some type of water vehicle involed with the incident. Next, I updated type1 again by grouping together Invalid, Questionable, Unconfirmed, Unverified, and Under Investigation and encompassing adn replacing those with Needs_examination. I did this because all of those cases seemed to be not fully understood or reliably documented. After, I dropped any possible na values from the newly grouped type1 data, put that in a new vector, and printed that new vector. I then rechecked if there were any null values as well as the unique value count to ensure that they changes I made were correctly done.

In [229]:
# Number 3
import numpy as np
import pandas as pd
justice1 = pd.read_parquet("/Users/borayadiul/Desktop/DS 3001/wrangling/data/justice_data.parquet", engine='pyarrow')
released1 = justice1['WhetherDefendantWasReleasedPretrial']
print(released1)
print(released1.value_counts())
print(sum(released1.isnull()))
released2 = released1.replace(9, np.nan)
print(released2.value_counts())

0        9
1        0
2        0
3        0
4        1
        ..
22981    1
22982    1
22983    1
22984    1
22985    1
Name: WhetherDefendantWasReleasedPretrial, Length: 22986, dtype: int64
WhetherDefendantWasReleasedPretrial
1    19154
0     3801
9       31
Name: count, dtype: int64
0
WhetherDefendantWasReleasedPretrial
1.0    19154
0.0     3801
Name: count, dtype: int64


For this question, since the data was in the format of parquet, I used pd.read_parquet in order to read the data in. I then created a new vector that only included the 'WhetherDefendantWasReleasedPretrial'column from the original dataset. Following that, I printed the new vector as well as print the unique value counts to get a sense of what the vector looks like. The value counts also helped me see which specific values were used to organize the data because I was then able to go to the codebook and see what each of the number meant for this dataset. Next, I printed if there are any null values within the released1 vector to see if those needed to be cleaned. After looking at the codebook, they coded 9 to mean 'Unclear', indicating that this data was could not reliable say whether someone had been released or not definitively. Further analysis of this data would probaly be focused on the clearly documented cases of whether someone was released (1) or not (0), therefore I decided to replace the obeservations that were coded as 9 with nan to inidcate that these values were 'missing' or undefined within the vector and store this cleaned data within a new vector. I did this to ensure that I still had copies of unedited data in cases there was a mistake. Finally, I printed out the unique value counts for the newly cleaned vector to make sure that the data was cleaned properly, the 9 values were taken out, and the original count for 1 and 0 remained unchanged.

In [227]:
# Number 4
import numpy as np
import pandas as pd
justice1 = pd.read_parquet("/Users/borayadiul/Desktop/DS 3001/wrangling/data/justice_data.parquet", engine='pyarrow')
imposed1 = justice1['ImposedSentenceAllChargeInContactEvent']
print(imposed1.value_counts())
print(sum(imposed1.isnull()))
sen_type1 = justice1["SentenceTypeAllChargesAtConvictionInContactEvent"]
sen_type1.value_counts()

imposed1 = pd.to_numeric(imposed1, errors='coerce')
imposed1_nan = imposed1.isnull()
print('Total Missings: \n', sum(imposed1_nan),'\n') 
print(pd.crosstab(imposed1_nan,sen_type1))

imposed1 = imposed1.mask(sen_type1 ==4, 0)
imposed1 = imposed1.mask(sen_type1 ==9, np.nan)

imposed1_nan = imposed1.isnull()
print(pd.crosstab(imposed1_nan, sen_type1))

print('Total Missings: \n', np.sum(imposed1_nan)) 


ImposedSentenceAllChargeInContactEvent
                    9053
0                   4953
12                  1404
.985626283367556    1051
6                    809
                    ... 
49.9712525667351       1
57.0349075975359       1
79.9260780287474       1
42.1642710472279       1
1.6570841889117        1
Name: count, Length: 484, dtype: int64
0
Total Missings: 
 9053 

SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
ImposedSentenceAllChargeInContactEvent                                      
False                                             8720  4299  914     0    0
True                                                 0     0    0  8779  274
SentenceTypeAllChargesAtConvictionInContactEvent     0     1    2     4    9
ImposedSentenceAllChargeInContactEvent                                      
False                                             8720  4299  914  8779    0
True                                                 0     0    0     0  274
Tota

For this question, I again imported the data using pd.read_parquet since the data was in a parquet format. Next, I created a new vector that only contained the 'ImposedSentenceAllChargeInContactEvent' column from the original dataset. I printed the unique value counts to see how the data looked like and I also printed the amount of null values that were present within the vector. In order to have more context behind what made up the total imposed sentence, I decided to also create a new vector that contained the 'SentenceTypeAllChargesAtConvictionInContactEvent' variable from the original dataset. The type of sentence one receives impacts the amount of time they may need to serve their sentences, as someone who was assigned a charge sentence that was later dropped will appear as missing in the imposed vector. I also did a unique value count of the sentence type vector to see the distribution, as well as to see what values to look up in the codebook to see what they indicate. Based on what was shown in the value counts for imposed1, I converted this vector into numeric values to get rid of any spaces and to make further analysis easier. After seeing that there were a large number of null, missing values, I placed those values into a dummy variable and printed out the total number of missing values to ensure it matched the amount of missing values I originally printed earlier. Then, I created a table looking at the missing values in the imposed vector alongside the sentence type. I did this to see which category of sentencing these missing values were falling into and to be able to see what value in the codebook that corresponded too. 
With true indicating the count of missing data and false indicating nonmissing data, the table revealed that 4 and 9 from the sentencing type vector contained the missing data. 4 indicates that the charges were pending, dimissed, or deffered and 9 indicates that the observation was no longer present. I decided to replace 4 with 0, to represent how those charges were not acutally missing data, and 9 with nan, to represent how those obersvations were missing. 
I then created another dummy variable for missing data to account for the changes in what should be counted as missing compared to what data is still present and created another table with this dummy variable and the sentence type to ensure that it properly documented that only 9 should have true values, indicating missing data. Finally, I printed out the total number of missing data in the dummy variable to make sure it matches what was in the table and different from the orignal number of 9053.

**Q3.** Many important datasets contain a race variable, typically limited to a handful of values often including Black, White, Asian, Latino, and Indigenous. This question looks at data gathering efforts on this variable by the U.S. Federal government.

1. How did the most recent US Census gather data on race?
2. Why do we gather these data? What role do these kinds of data play in politics and society? Why does data quality matter?
3. Please provide a constructive criticism of how the Census was conducted: What was done well? What do you think was missing? How should future large scale surveys be adjusted to best reflect the diversity of the population? Could some of the Census' good practices be adopted more widely to gather richer and more useful data?
4. How did the Census gather data on sex and gender? Please provide a similar constructive criticism of their practices.
5. When it comes to cleaning data, what concerns do you have about protected characteristics like sex, gender, sexual identity, or race? What challenges can you imagine arising when there are missing values? What good or bad practices might people adopt, and why?
6. Suppose someone invented an algorithm to impute values for protected characteristics like race, gender, sex, or sexuality. What kinds of concerns would you have?